gateway/ln-gateway/src/actor.rs

```diff
@@ -76,53 +81,80 @@ impl GatewayActor {
                     }
                 }
             }
-        });
+        })
+        .await;
 
-        let actor = Self {
+        let mut actor = Self {
             client,
             lnrpc,
-            task_group,
+            task_group: tg,
+            gw_rpc,
+            sender: None,
         };
 
         actor.subscribe_htlcs().await?;
 
         Ok(actor)
     }
 
-    async fn subscribe_htlcs(&self) -> Result<()> {
+    pub async fn stop_subscribing_htlcs(&mut self) -> Result<()> {
+        if let Some(sender) = &self.sender {
+            sender
+                .send(())
+                .await
+                .expect("Couldn't send signal to shutdown HTLC thread");
+        }
+        Ok(())
+    }
+
+    pub async fn subscribe_htlcs(&mut self) -> Result<()> {
         let short_channel_id = self.client.config().mint_channel_id;
-        let mut tg = self.task_group.clone();
+
+        // Create a channel that will be used to shutdown the HTLC thread
+        let (sender, mut receiver) = mpsc::channel::<()>(100);
+        self.sender = Some(sender);
 
         let mut stream = self
             .lnrpc
-            .to_owned()
+            .read()
+            .await
             .subscribe_htlcs(SubscribeInterceptHtlcsRequest { short_channel_id })
             .await?;
         info!("Subscribed to HTLCs with {:?}", short_channel_id);
 
         let actor = self.to_owned();
         let lnrpc_copy = self.lnrpc.to_owned();
-        tg.spawn(
+        let gw_rpc_copy = self.gw_rpc.clone();
+        self.task_group.spawn(
             "Subscribe to intercepted HTLCs in stream",
             move |subscription| async move {
                 while let Some(SubscribeInterceptHtlcsResponse {
                     payment_hash,
                     outgoing_amount_msat,
                     intercepted_htlc_id,
                     ..
-                }) = match stream.next().await {
-                    Some(msg) => match msg {
-                        Ok(msg) => Some(msg),
-                        Err(e) => {
-                            warn!("Error sent over HTLC subscription: {}", e);
+                }) = tokio::select! {
+                    msg = stream.next() => match msg {
+                        Some(msg) => match msg {
```

Comment:
right margin drift is very high here . Maybe move `while let Some(...)` to a sub-function?

Possibly

```
match msg {
  Some(Ok(msg)) => Some(msg),
  Some(Err(e))) => ...
...
}
```

just to remove one level of nesting?

Reply:
Moved to a function and removed that one level of nesting